### PR TITLE
Allow skipping action validation

### LINKED ISSUE #538: Collection modules inside tasks cause exceptions
## Issue
`ansible-lint` uses `ansible` `ModuleArgsParser` logic to extract the action from a task, it finds core modules successfully but starting with ansible 2.8 for collections it expects the collection to be loaded from the environment via other pieces of ansible, and a module from a collection causes an exception (related, #372 reports exception on custom modules)

## Background
Currently ansible-lint does not use other pieces of ansible to go outside the role it is linting and search for (and load) installed collections. This is probably best since there would be complexity and maintainability issues. Also if this were done, linting a role would yield different results depending on whether it was run inside an environment with installed collections, or standalone (at this point it may be violating the name “linter” by no longer being a static analysis tool)

## Proposal
Keep using ModuleArgsParser, catch a “no action detected” exception and then apply custom logic to do a best-guess of the action - choose the first argument that does not match a known keyword

### PR DESCRIPTION
**Pull request by ssbarnea**
To make way for modules called from a collection, ask Ansible to hold
off on parsing the module names. We cannot validate and guarantee,
successfully, that such modules are valid or not as part of a static
linting tool

Fixes #538

### CODE REVIEW

**Comment by ssbarnea** [comment]
AFAIK, there is another missing bit: that feature is not documented and currently listing tags via `-T` does not print any descriptions about tags.

Before merging this I will also have to refactor the way we print the tags in order to allow doc strings about them.

File: lib/ansiblelint/__main__.py [hunk 1]
```diff
@@ -125,8 +126,6 @@ def main() -> int:
     """Linter CLI entry point."""
     cwd = pathlib.Path.cwd()
 
-    options = cli.get_config(sys.argv[1:])
```
Comment (webknjaz) [inline lib/ansiblelint/__main__.py:128]: Removing this doesn't look right. Look for something that may be mutating this list (it's not immutable).

Comment (ssbarnea) [reply lib/ansiblelint/__main__.py:128]: @webknjaz we are not removing it, it is happening inside the `config` module. The only issue that we have is that  now we need options to be loaded inside `utils` module, and this affected test suites where we used to test different by just importing the methods we wanted to test. This now triggers config loading as side effect and fails because the command line arguments are not the same.

We could likely avoid it by mocking the import or mocking the cli arguments.

A longer term solution would be to move stuff outside the utils module which used to be the kitchensink.

Comment (webknjaz) [reply lib/ansiblelint/__main__.py:129]: That doesn't look right either. The typical idiom is to access `sys.argv` on the top level (that would be this module). The main function is supposed to accept it as an argument and the outer func would access `sys.argv` and do return `main(sys.argv)`. That outer func would also have to be specified as a console script entrypoint.

This way, anybody can cleanly call `main()` with any args they want from tests w/o having to worry about the state of the global `sys.argv`.

File: lib/ansiblelint/_internal/config.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+"""Loads config options."""
+import sys
+
+from ansiblelint import cli
+
+if any(i in ['pytest', 'sphinx'] for i in sys.modules):
+    options = cli.get_config([])
+else:
+    options = cli.get_config(sys.argv[1:])
```
Comment (webknjaz) [inline lib/ansiblelint/_internal/config.py:9]: This is a bad approach because it heavily relies on the global state of `sys.argv` and `sys.modules`. I don't see how this code can be treated as predictable.

**Comment by webknjaz** [comment]
@ssbarnea here's how the work with `sys.argv` is normally organized: https://github.com/ansible-community/ansible-lint/pull/1146

File: lib/ansiblelint/__main__.py (hunk 1 above)
Comment (ssbarnea) [reply lib/ansiblelint/__main__.py:129]: Please check again, I rewrote it.

File: lib/ansiblelint/__main__.py [hunk 3]
```diff
@@ -122,11 +123,15 @@ def report_outcome(matches: List["MatchError"], options) -> int:
...
-def main() -> int:
+def main(args=None) -> int:
     """Linter CLI entry point."""
+    if args is None:
+        args = sys.argv[1:]
+
     cwd = pathlib.Path.cwd()
 
-    options = cli.get_config(sys.argv[1:])
+    constants.options = get_config(args)
```
Comment (webknjaz) [inline lib/ansiblelint/__main__.py:133]: I don't think you can claim something to be constant if you literally rewrite its value spaghetti-style. Global vars that can be randomly modified from many places in the control flow are something to avoid at any cost.

File: lib/ansiblelint/utils.py [hunk 4]
```diff
@@ -57,7 +57,7 @@
 from yaml.representer import RepresenterError
 
 from ansiblelint._internal.rules import AnsibleParserErrorRule
-from ansiblelint.constants import CUSTOM_RULESDIR_ENVVAR, DEFAULT_RULESDIR, FileType
+from ansiblelint.constants import CUSTOM_RULESDIR_ENVVAR, DEFAULT_RULESDIR, FileType, options
```
Comment (webknjaz) [inline lib/ansiblelint/utils.py:60]: This is going to be problematic depending on whether `options` import happens before or after calling `main()` :exploding_head:

File: lib/ansiblelint/utils.py [hunk 5]
```diff
@@ -435,14 +435,16 @@ def _sanitize_task(task: dict) -> dict:
 
 # FIXME: drop noqa once this function is made simpler
 # Ref: https://github.com/ansible-community/ansible-lint/issues/744
-def normalize_task_v2(task: dict) -> dict:  # noqa: C901
+def normalize_task_v2(task: dict, options=options) -> dict:  # noqa: C901
```
Comment (webknjaz) [inline lib/ansiblelint/utils.py:438]: Using mutable objects as defaults is a known and dangerous antipattern that can lead to hard to follow behaviors.

**Comment by ssbarnea** [comment]
got too tired, closing for now.
